server/aesdsocket.c

```diff
@@ -0,0 +1,300 @@
+/*
+    Written by: Xavier Rojas
+
+*/
+#include <stdlib.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <sys/stat.h>
+#include <netinet/in.h>
+#include <arpa/inet.h>
+#include <netdb.h>
+#include <syslog.h>
+#include <signal.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <stdbool.h>
+#include <fcntl.h>
+
+#define PORT "9000"
+#define ERROR 0xFFFFFFFF
+#define BACKLOGS 5
+#define OUTPUT_FILE_PATH "/var/tmp/aesdsocketdata"
+
+volatile int sock_fd;
+volatile int signal_caught = 0;
+
+static void signal_handler(int sig)
+{
+    // Just shutdown, cleanup happens in main
+    shutdown(sock_fd, SHUT_RDWR);
+    signal_caught = sig;
+}
+
+// // Guided by https://www.thegeekstuff.com/2012/02/c-daemon-process/
+static void start_daemon()
+{
+    closelog();
+    openlog("AESD Socket Daemon Log", 0, LOG_USER);
+    pid_t process_id = 0;
+    pid_t sid = 0;
+
+    // Child process
+    process_id = fork();
+    if (process_id < 0)
+    {
+        fprintf(stderr, "ERROR, could not fork\n");
+        close(sock_fd);
+        exit(ERROR);
+    }
+
+    // Kill parent
+    if (process_id > 0)
+    {
+        syslog(LOG_INFO, "Killing parent process for daemon setup");
+        exit(0);
+    }
+
+    // Set permissions
+    umask(0);
+
+    // New session
+    sid = setsid();
+    if (sid < 0)
+    {
+        fprintf(stderr, "ERROR, could not create new session\n");
+        close(sock_fd);
+        exit(ERROR);
+    }
+
+    // Change wd to root
+    chdir("/");
+
+    // Close pertinent files
+    close(STDIN_FILENO);
+    close(STDOUT_FILENO);
+    close(STDERR_FILENO);
+    // Redirect, got this from: https://groups.google.com/g/comp.unix.programmer/c/cJJoGqHmztI
+    open("/dev/null", O_RDWR);
+    dup(0);
+    dup(0);
+
+    syslog(LOG_INFO, "Daemon initialized");
+}
+
+// Ideas pulled from https://github.com/jasujm/apparatus-examples/blob/master/signal-handling/lib.c
+void transaction(int client_fd)
+{
+    FILE *output_file;
+    int nbytes_buf = 512;
+    int nbytes_recv;
+    char* buf_end = NULL;
+    char recv_buf[nbytes_buf];
+
+    // Open file outside of the loop
+    output_file = fopen(OUTPUT_FILE_PATH, "a+");
+    if (!output_file)
+    {
+        fprintf(stderr, "ERROR opening output file for writing\n");
+        exit(ERROR);
+    }
+
+    // Read message into buffer
+    while (!buf_end)
+    {
+        // Alwas reset the buf no matter what
+        memset(recv_buf, 0, sizeof(recv_buf));
+        if ((nbytes_recv = recv(client_fd, recv_buf, sizeof(recv_buf), 0)) < 0)
+        {
+            fprintf(stderr, "ERROR calling recv\n");
+            exit(ERROR);
+        }
+
+        // Determine if end of packet
+        buf_end = (char *)memchr(recv_buf, '\n', nbytes_recv);
+
+        // If we never detected a newline, just append without a newline
+        fwrite(recv_buf, nbytes_recv, 1, output_file);
```

Comment:
@rojasx this will write to the file before the packet completes, which may cause problems with future assignments.